java/dagger/internal/codegen/BindingGraph.java

```diff
@@ -1120,6 +1118,23 @@ private boolean hasLocallyPresentOptionalBinding(ResolvedBindings resolvedBindin
               && !getLocalExplicitBindings(keyFactory.unwrapOptional(resolvedBindings.key()).get())
                   .isEmpty();
         }
+
+        /**
+         * Returns {@code true} if {@code resolvedBindings} contains a binding for which there is
+         * an explicit present binding in this component that was resolved previously.
+         */
+        boolean hasLocalContributionBinding(ResolvedBindings resolvedBindings) {
+          return resolvedBindings
+                .contributionBindings()
+                .stream()
+                .anyMatch(contributionBinding ->
+                    !getLocalExplicitBindings(resolvedBindings.key()).isEmpty()
+                  && parentResolver.isPresent()
+                  && parentResolver.get().getPreviouslyResolvedBindings(
+                      resolvedBindings.bindingKey()).isPresent()
+                  && !parentResolver.get().getPreviouslyResolvedBindings(
```

Comment:
You're right, this method shouldn't care if the parent bindings are empty. I've removed this check in 4f78aab. 

However, in addition to the case you mentioned above, `LocalDependencyChecker` is also used to determine the owning component of a binding in `getOwningComponent()`. Now that we're checking for local contribution bindings here, we incorrectly determine the owning component of the `Object` binding to be `Child`.

To address this I think we need to add some additional logic to `getOwningComponent()`. Does this make sense?

```
@Module
class ParentModule { 
  @Provides Object needsString(String string) { return new Object; } 
}

@Component(modules = ParentModule.class)
interface Parent { Child child(); }

@Module
class ChildModule {
  @Provides String string() { return "Hi!" }
}

@Subcomponent(modules = ChildModule.class)
interface Child {
  String string();
  Object object();
}
```